File: e2e/fsm-wake-and-skip-test.py
#!/usr/bin/env python3
"""Follow-up unit tests for two FSM bugs found on the real Shield. No Plex, no MQTT, no ADB,
no network — the adb / profiles / playback primitives are stubbed.

Bug 1 — dozing Shield not woken before launch. `adb.ensure_plex_open` did `am start plex://`
        on a sleeping device, which does NOT foreground Plex, so the launch failed until
        something else sent a WAKEUP. FIX: the device_on transition — wake FIRST when the
        Shield isn't awake (or the foreground is unknown), then launch.

Bug 2 — picker walked even when already on the required profile. `_drive_profile`'s
        "already on required -> skip" never had data: the FSM gated path never populated
        `profiles.LAST_SEEN`, so the alias-aware skip couldn't fire and the picker was walked
        on every gated scan. FIX: an alias-aware skip (display-name 'Bob Smith' ==
        username 'sawtaytoes') AND recording the confirmed profile into LAST_SEEN after a
        switch, so subsequent gated scans short-circuit with NO picker.

Run:  python3 e2e/fsm-wake-and-skip-test.py     (from the repo root; exits non-zero on failure)
"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from queue_builder import adb, config, driver, profiles  # noqa: E402

FAILS = []


def ok(name, cond, detail=""):
    print(("PASS " if cond else "FAIL ") + name + (f"  -- {detail}" if detail and not cond else ""))
    if not cond:
        FAILS.append(name)


# =========================================================================== #
# Bug 1 — ensure_plex_open wakes a dozing Shield BEFORE launching Plex
# =========================================================================== #
# Record every low-level adb call in order so we can assert WAKEUP precedes `am start`.
def make_adb_env(awake, foreground_sequence):
    order = []
    seq = iter(foreground_sequence)
    last = {"v": foreground_sequence[-1] if foreground_sequence else None}

    def _run(args, timeout=None):
        if args[:2] == ["shell", "am"] or (len(args) > 1 and args[0] == "shell"
                                           and args[1].startswith("am ")):
            pass
        # am start plex://
        if args[0] == "shell" and "am" in args and "start" in args:
            order.append("launch")
            return ""  # non-None -> "issued"
        return ""

    def foreground_activity():
        try:
            last["v"] = next(seq)
        except StopIteration:
            pass
        return last["v"]

    def is_awake():
        return awake

    def _press(*keycodes, settle=0.0):
        if "KEYCODE_WAKEUP" in keycodes:
            order.append("wake")
        return True

    adb.connect = lambda: True
    adb._run = _run
    adb.foreground_activity = foreground_activity
    adb.is_awake = is_awake
    adb._press = _press
    adb._WAKE_SETTLE = 0  # no real sleep in the test
    config.ADB_PLEX_LAUNCH_WAIT_SECONDS = 2
    return order


# Asleep/dozing: foreground reads null, then null after wake, then Plex once launched.
_PLEX = "com.plexapp.android/.HomeActivityTV"
order = make_adb_env(awake=False, foreground_sequence=[None, None, _PLEX])
res = adb.ensure_plex_open()
ok("(bug1) asleep: WAKEUP is sent", "wake" in order)
ok("(bug1) asleep: launch happens", "launch" in order)
ok("(bug1) asleep: WAKEUP is sent BEFORE the plex:// launch",
   "wake" in order and "launch" in order and order.index("wake") < order.index("launch"),
   str(order))
ok("(bug1) asleep: ends foreground on Plex -> True", res is True)

# Wake alone brings Plex back (screensaver was over Plex): no launch needed.
order = make_adb_env(awake=False, foreground_sequence=[None, _PLEX])
res = adb.ensure_plex_open()
ok("(bug1) asleep-over-plex: wakes and returns True without launching",
   "wake" in order and "launch" not in order and res is True, str(order))

# Awake on the launcher: no needless WAKEUP, straight to launch.
order = make_adb_env(awake=True, foreground_sequence=["com.google.android.tvlauncher/.X", _PLEX])
res = adb.ensure_plex_open()
ok("(bug1) awake-on-launcher: does NOT send a needless WAKEUP", "wake" not in order, str(order))
ok("(bug1) awake-on-launcher: launches and returns True", "launch" in order and res is True)

# Already on Plex: returns immediately, no wake, no launch.
order = make_adb_env(awake=True, foreground_sequence=[_PLEX])
res = adb.ensure_plex_open()
ok("(bug1) already-on-plex: no wake, no launch, True",
   order == [] and res is True, str(order))


# =========================================================================== #
# Bug 2 — _drive_profile skips the picker when already on `required` (alias-aware),
#         and records the confirmed profile so the NEXT scan skips too.
# =========================================================================== #
SWITCHES = []


def wire_driver(alias_groups=(), switch_ok=True):
    """Stub adb.same_profile / adb.switch_to for the driver; count switch_to calls."""
    SWITCHES.clear()
    config.ADB_ENABLED = True
    config.PLAYBACK_FSM_SWITCH_ATTEMPTS = 2
    config.PLAYBACK_FSM_RETRY_BACKOFF = 0

    groups = [set(g) for g in alias_groups]

    def same_profile(a, b):
        if a == b:
            return True
        for g in groups:
            if a in g and b in g:
                return True
        return False

    def switch_to(target, cancel=None, known_current=None):
        SWITCHES.append((target, known_current))
        return (switch_ok, "selected on the picker" if switch_ok else "ran out of time")

    adb.same_profile = same_profile
    adb.switch_to = switch_to


# (a) LAST_SEEN == required exactly -> no picker walk.
wire_driver()
profiles.LAST_SEEN["title"] = "sawtaytoes"
r = driver._drive_profile(None, "sawtaytoes", None)
ok("(bug2a) exact LAST_SEEN==required: gate satisfied (None)", r is None)
ok("(bug2a) exact LAST_SEEN==required: NO switch_to call", len(SWITCHES) == 0, str(SWITCHES))

# (b) LAST_SEEN is the DISPLAY name, required the USERNAME — alias must short-circuit.
wire_driver(alias_groups=[{"Bob Smith", "sawtaytoes"}])
profiles.LAST_SEEN["title"] = "Bob Smith"
r = driver._drive_profile(None, "sawtaytoes", None)
ok("(bug2b) display-name==username alias: gate satisfied (None)", r is None)
ok("(bug2b) display-name==username alias: NO picker/switch call", len(SWITCHES) == 0, str(SWITCHES))

# (c) A real change: LAST_SEEN cold (None) -> switch once, and LAST_SEEN is then RECORDED as
#     `required` so an immediate second call short-circuits with no further switch.
wire_driver(alias_groups=[{"Bob Smith", "sawtaytoes"}], switch_ok=True)
profiles.LAST_SEEN["title"] = None
r1 = driver._drive_profile(None, "sawtaytoes", None)
ok("(bug2c) cold cache: switch runs once", r1 is None and len(SWITCHES) == 1, str(SWITCHES))
ok("(bug2c) switch records the profile into LAST_SEEN",
   profiles.LAST_SEEN.get("title") == "sawtaytoes", str(profiles.LAST_SEEN))
r2 = driver._drive_profile(None, "sawtaytoes", None)
ok("(bug2c) second gated scan short-circuits: still only ONE switch total",
   r2 is None and len(SWITCHES) == 1, str(SWITCHES))

# (d) Cold cache + a DIFFERENT signed-in profile -> a real switch is still driven.
wire_driver(alias_groups=[{"Bob Smith", "sawtaytoes"}, {"Younger Kids"}])
profiles.LAST_SEEN["title"] = "Younger Kids"  # signed in as someone else
r = driver._drive_profile(None, "sawtaytoes", None)
ok("(bug2d) genuinely-wrong profile: drives the switch once", r is None and len(SWITCHES) == 1,
   str(SWITCHES))


print(f"\nFAILURES: {len(FAILS)}" if FAILS else "\ndone")
sys.exit(1 if FAILS else 0)

File: e2e/fsm-wake-and-skip-test.ts
// Unit tests for two FSM bugs found on the real Shield. No Plex, no MQTT, no ADB, no network.
//
// Node port of the retired e2e/fsm-wake-and-skip-test.py (deleted with queue_builder/ on
// 2026-08-12), asserting the same two fixes against server/src/adb.js + server/src/driver.js:
//
// Bug 1 — dozing Shield not woken before launch. `am start plex://` on a sleeping device does
//         NOT foreground Plex — the launch queues behind the dream. FIX: wake FIRST when the
//         Shield isn't awake (or the foreground is unknown), then launch. Driven through the
//         REAL adb.js with node:child_process scripted (e2e/stubs/child-process.stub.mjs), so
//         the wake→launch ORDER is observed, not assumed.
//
// Bug 2 — picker walked even when already on the required profile. driveProfile's
//         "already on required -> skip" needs profiles.LAST_SEEN populated and alias-aware
//         matching; a successful switch must RECORD the profile so the next scan skips.
//
// Run:  server/node_modules/.bin/tsx e2e/fsm-wake-and-skip-test.ts   (from the repo root; exits non-zero on failure)

process.env.ADB_ENABLED = 'true';
process.env.ADB_PLEX_LAUNCH_WAIT_SECONDS = '1'; // bound the post-launch poll in tests
process.env.PLAYBACK_FSM_SWITCH_ATTEMPTS = '2';
process.env.PLAYBACK_FSM_RETRY_BACKOFF = '0';

import { stubAdbShell, stubDriverDeps } from './stubs/hooks.mjs';

stubAdbShell();
stubDriverDeps();
const {
  ADB_SCRIPT: RAW_ADB_SCRIPT, resetScript: RAW_RESET_SCRIPT,
} = await import('./stubs/child-process.stub.mjs');
const { CTL: RAW_CTL, reset } = await import('./stubs/control.mjs');
const adb = await import('../server/src/adb.js'); // REAL module, faked shell
const driver = await import('../server/src/driver.js'); // stubbed adb/playback/profiles

/**
 * The two stub control surfaces, typed.
 *
 * `stubs/child-process.stub.mjs` and `stubs/control.mjs` stay hand-written JavaScript — they
 * are the `registerHooks` machinery and the mutable object its stubs close over — so under
 * `allowJs`/`checkJs: false` both are INFERRED from initializers of `[]`/`null`: `order` and
 * `foreground` come out `never[]`, `lastSeen.title` the type `null`. Scripting a scenario is
 * then an error about the fixture rather than about the FSM. These declare what the stubs
 * actually carry at runtime; they are views of the same objects, not copies.
 */
interface AdbScript {
  /** 'wake' / 'launch', in the order adb.js issued them. */
  order: string[];
  awake: boolean;
  /** Queue of foreground activities; the last one repeats once drained. */
  foreground: (string | null)[];
  _lastForeground: string | null;
}
interface DriverCtl {
  /** `[name, ...args]` per recorded primitive — read positionally by the assertions. */
  calls: unknown[][];
  foreground: string;
  companionUp: boolean;
  switchResult: [boolean, string];
  /** `"a|b"` -> the sameProfile answer to force. */
  same: Map<string, boolean>;
  playResults: unknown[];
  lastSeen: { title: string | null };
  awake: boolean;
  onSwitch: (() => void) | null;
}
const ADB_SCRIPT = RAW_ADB_SCRIPT as unknown as AdbScript;
const resetScript = RAW_RESET_SCRIPT as unknown as
  (opts?: { awake?: boolean; foreground?: (string | null)[] }) => void;
const CTL = RAW_CTL as unknown as DriverCtl;

const FAILS: string[] = [];
function ok(name: string, cond: boolean, detail = ''): void {
  console.log((cond ? 'PASS ' : 'FAIL ') + name + (!cond && detail ? `  -- ${detail}` : ''));
  if (!cond) FAILS.push(name);
}

// =========================================================================== //
// Bug 1 — ensurePlexOpen wakes a dozing Shield BEFORE launching Plex
// =========================================================================== //
const PLEX = 'com.plexapp.android/.HomeActivityTV';
const LAUNCHER = 'com.google.android.tvlauncher/.MainActivity';

// Asleep/dozing: foreground reads null, then null after the wake, then Plex once launched.
resetScript({ awake: false, foreground: [null, null, PLEX] });
let res = await adb.ensurePlexOpen();
ok('(bug1) asleep: WAKEUP is sent', ADB_SCRIPT.order.includes('wake'));
ok('(bug1) asleep: launch happens', ADB_SCRIPT.order.includes('launch'));
ok(
  '(bug1) asleep: WAKEUP is sent BEFORE the plex:// launch',
  ADB_SCRIPT.order.indexOf('wake') >= 0
    && ADB_SCRIPT.order.indexOf('wake') < ADB_SCRIPT.order.indexOf('launch'),
  JSON.stringify(ADB_SCRIPT.order),
);
ok('(bug1) asleep: ends foreground on Plex -> true', res === true);

// Wake alone brings Plex back (the screensaver was over Plex): no launch needed.
resetScript({ awake: false, foreground: [null, PLEX] });
res = await adb.ensurePlexOpen();
ok(
  '(bug1) asleep-over-plex: wakes and returns true without launching',
  ADB_SCRIPT.order.includes('wake') && !ADB_SCRIPT.order.includes('launch') && res === true,
  JSON.stringify(ADB_SCRIPT.order),
);

// Awake on the launcher: no needless WAKEUP, straight to launch.
resetScript({ awake: true, foreground: [LAUNCHER, PLEX] });
res = await adb.ensurePlexOpen();
ok('(bug1) awake-on-launcher: does NOT send a needless WAKEUP',
  !ADB_SCRIPT.order.includes('wake'), JSON.stringify(ADB_SCRIPT.order));
ok('(bug1) awake-on-launcher: launches and returns true',
  ADB_SCRIPT.order.includes('launch') && res === true);

// Already on Plex: returns immediately, no wake, no launch.
resetScript({ awake: true, foreground: [PLEX] });
res = await adb.ensurePlexOpen();
ok('(bug1) already-on-plex: no wake, no launch, true',
  ADB_SCRIPT.order.length === 0 && res === true, JSON.stringify(ADB_SCRIPT.order));

// =========================================================================== //
// Bug 2 — driveProfile skips the picker when already on `required` (alias-aware),
//         and records the confirmed profile so the NEXT scan skips too.
// =========================================================================== //
const { driveProfile } = driver._internals;

// Wire the stubbed adb.sameProfile with alias groups, and count switchTo calls.
function wireDriver(aliasGroups: readonly (readonly string[])[] = [], switchOk = true): void {
  reset();
  CTL.switchResult = [switchOk, switchOk ? 'selected on the picker' : 'ran out of time'];
  for (const group of aliasGroups) {
    for (const a of group) {
      for (const b of group) CTL.same.set(`${a}|${b}`, true);
    }
  }
}
const switches = () => CTL.calls.filter((c) => c[0] === 'switch_to');

// (a) LAST_SEEN == required exactly -> no picker walk.
wireDriver();
CTL.lastSeen.title = 'sawtaytoes';
let r = await driveProfile(null, 'sawtaytoes', null);
ok('(bug2a) exact LAST_SEEN==required: gate satisfied (null)', r === null);
ok('(bug2a) exact LAST_SEEN==required: NO switchTo call', switches().length === 0);

// (b) LAST_SEEN is the DISPLAY name, required the USERNAME — the alias must short-circuit.
wireDriver([['Bob Smith', 'sawtaytoes']]);
CTL.lastSeen.title = 'Bob Smith';
r = await driveProfile(null, 'sawtaytoes', null);
ok('(bug2b) display-name==username alias: gate satisfied (null)', r === null);
ok('(bug2b) display-name==username alias: NO picker/switch call', switches().length === 0);

// (c) A real change: LAST_SEEN cold (null) -> switch once, and LAST_SEEN is then RECORDED as
//     `required`, so an immediate second call short-circuits with no further switch.
wireDriver([['Bob Smith', 'sawtaytoes']], true);
CTL.lastSeen.title = null;
const r1 = await driveProfile(null, 'sawtaytoes', null);
ok('(bug2c) cold cache: switch runs once', r1 === null && switches().length === 1,
  JSON.stringify(switches()));
ok('(bug2c) switch records the profile into LAST_SEEN', CTL.lastSeen.title === 'sawtaytoes',
  String(CTL.lastSeen.title));
const r2 = await driveProfile(null, 'sawtaytoes', null);
ok('(bug2c) second gated scan short-circuits: still only ONE switch total',
  r2 === null && switches().length === 1, JSON.stringify(switches()));

// (d) Cold cache + a DIFFERENT signed-in profile -> a real switch is still driven.
wireDriver([['Bob Smith', 'sawtaytoes'], ['Younger Kids']]);
CTL.lastSeen.title = 'Younger Kids'; // signed in as someone else
r = await driveProfile(null, 'sawtaytoes', null);
ok('(bug2d) genuinely-wrong profile: drives the switch once',
  r === null && switches().length === 1, JSON.stringify(switches()));

console.log(FAILS.length ? `\nFAILURES: ${FAILS.length}` : '\ndone');
process.exit(FAILS.length ? 1 : 0);
